File: static/javascripts/transaction.js
/*
Copyright Fujitsu Corp. All Rights Reserved.

SPDX-License-Identifier: Apache-2.0
*/
$(function () {

  $("input").val("")
  $("#select-paging").val(10)

  // get channelid
  $.ajax({
    url: channelIdUrl,
    timeout: 3000
  })
    .done(function (data) {
      var getData,
        channelID,
        transactionParam,
        transactionBlockNumberName;
      if (typeof (data) == "string") {
        getData = JSON.parse(data);
      } else if (typeof (data) == "object") {
        getData = data;
      }
      if (getData.code !== 200) {
        alert(getData.msg)
        return false;
      }
      channelID = getData.data[0].id

      // get the data of the Creator of tx
      creatorOftxData(transactionCreateUrl + channelID)

      transactionParam = location.search
      transactionBlockNumberName = transactionParam.slice((transactionParam.indexOf("=") + 1), transactionParam.length)

      //check if url parameter exists
      if (transactionBlockNumberName !== "") {  // has parameters

        var $tableTh = $("#block-table thead tr th");
        $tableTh.eq(0).removeClass("sorting_desc")
        $tableTh.eq(1).addClass("sorting_desc")
        $("#block-number-input").val(transactionBlockNumberName)
        $("#tab-click").find("li").eq(1).addClass("blocks-active").siblings().removeClass("blocks-active")
        $("#table-search>.form-inline").eq(1).removeClass("display-none").siblings().addClass("display-none")

        $("#paging-length, #entries-info, #entries-page").removeClass("display-none")
        initTableData(transactionBlockNumberUrl, channelID, "transactionBlockNumber")

      } else {                                      //no parameters
        //initialize the table of Time Range
        initTableData(transactionTimeUrl, channelID, "transactionTime")
      }

      // click the tab element, select the corresponding content part.
      $("#tab-click").find("li").on("click", debounce(function () {
        var $thisIndex = $(this).index(),
          $tableThtrFirst = $("#block-table thead tr th:eq(0)"),
          $tableThtrSecond = $("#block-table thead tr th:eq(1)"),
          $pageElement = $("#paging-length,#entries-info,#entries-page");

        $("#block-load").remove();
        $("#table-search").addClass("animated flipInX")
        setTimeout(function () { $("#table-search").removeClass("animated flipInX") }, 1000)
        $("#table-search>.form-inline").eq($thisIndex).removeClass("display-none").siblings().addClass("display-none")
        $(this).addClass("blocks-active").siblings().removeClass("blocks-active")

        $("#start-time,#end-time,#block-hash-input").val("")
        $("#select-paging").val(10)
        $("#block-table tbody tr").remove()
        $tableThtrFirst.removeClass("sorting_desc sorting_asc")
        $tableThtrSecond.removeClass("sorting_desc sorting_asc")

        if ($thisIndex == 0) {                        //Time Range part

          $tableThtrFirst.addClass("sorting_desc")
          $pageElement.removeClass("display-none")

          initTableData(transactionTimeUrl, channelID, "transactionTime")

        } else if ($thisIndex == 1) {             //Block Height part

          //   $tableThtrSecond.addClass("sorting_desc")
          $("#block-table").append('<tr><td class="text-center" colspan="8">Input Block Height to Query Tx Please...</td></tr>')
          //no parameters

          $("#block-number-input").val("")
          $pageElement.addClass("display-none")

        } else if ($thisIndex == 2) {                //Tx Hash part

          $("#block-hash-input").val("")
          $pageElement.addClass("display-none")
          $("#block-table").append('<tr><td class="text-center" colspan="8">Input Tx Hash to Query Tx Please...</td></tr>')

        }
      }))

      // search by Time Range
      $("#block-time-search").on("click", debounce(function () {
        initTableData(transactionTimeUrl, channelID, "transactionTime")
      }))

      // search by Blcok Height
      $("#block-number-search").off("click").on("click", debounce(function () {
        var reg = /^\d+$/g
        var $input = $("#block-number-input").val().trim();

        if (!reg.test($input)) {
          alert("Please fill in non-negative integers!")
          return false;
        }
        $("#block-table thead tr th:eq(1)").addClass("sorting_desc")
        $("#paging-length,#entries-info,#entries-page").removeClass("display-none")
        initTableData(transactionBlockNumberUrl, channelID, "transactionBlockNumber")

      }))

      // search by Data Hash
      $("#block-hash-search").off("click").on("click", debounce(function () {
        var $input = $("#block-hash-input").val().trim();
        if ($input == "") {
          alert("Cannot be empty!")
          return false;
        }

        txHashTableData(transactionHashUrl + channelID + "/" + $input)

      }))

      //  select the data, show the number of informations on each page.
      $("#select-paging").on("change", function () {
        var $tabTxt = $("#tab-click .blocks-active").index(),
          postUrl,
          choiceTxt;

        if ($tabTxt == 0) {  // Time Range Part

          postUrl = transactionTimeUrl
          choiceTxt = "transactionTime"

        } else if ($tabTxt == 1) {  // Block Height Part

          postUrl = transactionBlockNumberUrl
          choiceTxt = "transactionBlockNumber"

        }

        initTableData(postUrl, channelID, choiceTxt)

      })

      // the part of Time Range sort by Block Height
      // the part of Block Height sort by Tx Number
      $("#block-table thead tr th:eq(0),#block-table thead tr th:eq(1)").on("click", debounce(function () {

        var $self = $(this);
        ($self.hasClass("sorting_desc")) ? $self.removeClass("sorting_desc").addClass("sorting_asc") : $self.removeClass("sorting_asc").addClass("sorting_desc")

        var postUrl,
          choiceTxt,
          $tabTxt = $("#tab-click .blocks-active").index();

        if ($tabTxt == 0) {  // Time Range part

          postUrl = transactionTimeUrl
          choiceTxt = "transactionTime"

        } else if ($tabTxt == 1) {  // Block Height part

          postUrl = transactionBlockNumberUrl
          choiceTxt = "transactionBlockNumber"

        }

        initTableData(postUrl, channelID, choiceTxt)

      }))

      // click the word 'datahash' in the table, show some details
      $("#block-table").off("click").on("click", ".data-hash", debounce(function () {

        $('#read-set, #write-set, #endorser-signature').empty()

        var transactionParams = {}
        transactionParams.channelId = channelID
        transactionParams.txHash = $(this).text()
        //get data of tx hash
        txHashDetails(transactionHashDetailsUrl + channelID + "/" + $(this).text(), transactionParams)

      }))

      // click on the icon to implement the copy function
      copy($("#block-form2"), ".fa-copy")
    })
    .fail(function (jqXHR, textStatus) {
      networkStatus(jqXHR.status)
      alert("Request failed: " + textStatus);
    });



})

/*
* Initialize the table content.
* The table of Time Range , table codes.
* use AJAX to get the data.
*/

function initTableData(url, channelID, choiceTxt) {
  $("#block-load").remove();
  loading($(".block-table-contain"))

  var postParams = {},
    sizePageParam = Number($("#select-paging").val());

  if (choiceTxt == "transactionBlockNumber") {
    postParams.channel_id = channelID
    postParams.page_num = 1
    postParams.page_count = sizePageParam
    postParams.block_num = Number($("#block-number-input").val().trim())
    postParams.sort_by = $("#block-table tr th:eq(1)").hasClass("sorting_desc") ? "DESC" : "ASC"

  } else if (choiceTxt == "transactionTime") {

    postParams.channel_id = channelID
    postParams.sort_by = $("#block-table tr th:eq(0)").hasClass("sorting_desc") ? "DESC" : "ASC"
    postParams.start_time = $("#start-time").val().trim()
    postParams.end_time = $("#end-time").val().trim()
    postParams.page_num = 1
    postParams.page_count = sizePageParam
    postParams.create_msp = $("#transaction-creater").val() == "Please select" ? "" : $("#transaction-creater").val()

  }

  $.ajax({
    url: url,
    method: "POST",
    data: JSON.stringify(postParams),
    contentType: 'application/json',
    dataType: "json",
  })
    .done(function (data) {
      $("#block-load").remove();
      $("#block-table tbody tr").remove()
      // table 
      var getData,
        tableLen,
        tableData;
      if (typeof (data) == "string") {
        getData = JSON.parse(data);
      } else if (typeof (data) == "object") {
        getData = data;
      }
      if (getData.code !== 200) {
        $("#block-table").append('<tr><td class="text-center" colspan="8">' + getData.msg + '</td></tr>')
        return false;
      }

      tableLen = getData.data.count;
      tableData = getData.data.txs;
      $("#block-table").createTable({ tableData: tableData, tableLength: tableLen, tableId: "#block-table", pageId: "#table-paginate", tableInfoId: "#table-info", sizePage: sizePageParam, currentPage: 1, channelId: channelID, choiceContent: choiceTxt, previewUrl: url })

    })
    .fail(function (jqXHR, textStatus) {
      networkStatus(jqXHR.status)
      $("#block-load .loading").empty()
      $("#block-load .loading").append('<div class="text-center error-color">Error: Check Your Inputs Please， And Refresh This Page!</div>')
    });

}

// get the data of creator of tx
function creatorOftxData(url) {
  $.ajax({
    url: url,
  })
    .done(function (data) {
      var getData,
        arrayData,
        getDataLen,
        strHtml = "";

      if (typeof (data) == "string") {
        getData = JSON.parse(data);
      } else if (typeof (data) == "object") {
        getData = data;
      }
      $("#transaction-creater option:not(:first-child)").remove()

      if (getData.code == 200) {
        arrayData = getData.data
        getDataLen = arrayData.length;

        if (!!getDataLen) {
          for (var i = 0; i < getDataLen; i++) {
            strHtml += '<option value="' + arrayData[i] + '">' + arrayData[i] + '</option>'
          }
          $("#transaction-creater").append(strHtml)
        }
      } else {
        alert(getData.msg)
      }

    })
    .fail(function (jqXHR, textStatus) {
      networkStatus(jqXHR.status)
      alert("Request failed: " + textStatus);
    });

}
// get the data of tx hash
function txHashTableData(url) {
  loading($(".block-table-contain"))
  $.ajax({
    url: url,
  })
    .done(function (data) {
      $("#block-load").remove();
      $("#block-table tbody tr").remove()
      var getData;

      if (typeof (data) == "string") {
        getData = JSON.parse(data);
      } else if (typeof (data) == "object") {
        getData = data;
      }
      if (getData.code !== 200) {
        $("#block-table").append('<tr><td class="text-center" colspan="8">' + getData.msg + '</td></tr>')
        return false;
      }

      getData = getData.data;
      if (!!getData) {
        $("#block-table").append('<tr class="odd"><td>' + getData.block_num + '</td><td>' + getData.tx_num + '</td><td class="block-name"><div class="data-hash" data-toggle="modal" data-target="#myModal2" title="' + getData["tx_hash"] + '">' + getData["tx_hash"] + '</div></td><td><div class="block-hash">' + getData.chaincode_name + '</div></td><td><div class="previous-hash">' + getData.tx_type + '</div></td><td>' + getData.create_msp_id + '</td><td>' + getData.endorser_msp_id.replace(/,/g, '/') + '</td><td>' + getData.created_at.slice(0, 19).replace(/T/g, " ") + '</td></tr>');
      } else {
        $("#block-table").append('<tr><td class="text-center" colspan="8">find no transaction...</td></tr>')
      }

    })
    .fail(function (jqXHR, textStatus) {
      networkStatus(jqXHR.status)
      $("#block-load .loading").empty()
      $("#block-load .loading").append('<div class="text-center error-color">Error: Check Your Inputs Please， And Refresh This Page!</div>')
    });

}
// Tx Hash details
function txHashDetails(url) {
  $.ajax({
    url: url,
  })
    .done(function (data) {
      var getData;

      if (typeof (data) == "string") {
        getData = JSON.parse(data);
      } else if (typeof (data) == "object") {
        getData = data;
      }
      getData = getData.data
      var $form = $("#block-form2 .block-detail span")

      $form.eq(0).text(getData.block_num)
      $form.eq(1).text(getData.tx_num)
      $form.eq(2).text(getData.tx_hash)
      $form.eq(3).text(getData.chaincode_name)
      $form.eq(4).text(getData.tx_type)
      $form.eq(5).text(getData.chaincode_function)
      $form.eq(6).text(getData.chaincode_function_parameters)

      $form.eq(7).text(getData.create_msp_id)
      $form.eq(8).text(getData.endorser_msp_id.replace(/,/g, '/'))
      $form.eq(9).text(getData.created_at.slice(0, 19).replace(/T/g, " "))

      var options = {
        collapsed: false,
        withQuotes: true
      };

      $('#read-set').jsonViewer(JSON.parse(getData.read_set), options);

      var writeArr = JSON.parse(getData.write_set)
      if (!writeArr) {
        var writeArrLen = writeArr.length,
          writeSetArr,
          writeSetArrLen;

        for (var i = 0; i < writeArrLen; i++) {
          writeSetArr = writeArr[i]["set"];
          writeSetArrLen = writeSetArr ? writeSetArr.length : 0;
          for (var j = 0; j < writeSetArrLen; j++) {
            writeSetArr[j]["value"] = JSON.parse(writeSetArr[j]["value"])
          }
        }

      }

      $('#write-set').jsonViewer(writeArr, options);

      $("#endorser-signature").jsonViewer(JSON.parse(getData.endorser_signature), options);
    })
    .fail(function (jqXHR, textStatus) {
      networkStatus(jqXHR.status)
      alert("Request failed: " + textStatus);
    });
}

File: model/transaction.go
/*
Copyright Fujitsu Corp. All Rights Reserved.

SPDX-License-Identifier: Apache-2.0
*/
package model

import (
	"github.com/lib/pq"
	"github.com/pkg/errors"
	"gorm.io/gorm"
	"time"
)

type Transaction struct {
	ID                          int            `json:"id" gorm:"primarykey"`
	ChannelId                   int            `json:"channel_id" gorm:"type:integer"`
	BlockNum                    uint64         `json:"block_num" gorm:"type:integer"`
	TxNum                       uint64         `json:"tx_num" gorm:"type:integer"`
	TxHash                      string         `json:"tx_hash" gorm:"type:varchar(256)"`
	ChaincodeName               string         `json:"chaincode_name" gorm:"type:varchar(256)"`
	TxType                      string         `json:"tx_type" gorm:"type:varchar(256)"`
	TxFilter                    int            `json:"tx_filter" gorm:"type:integer"`
	TxResponse                  int32          `json:"tx_response" gorm:"type:integer"`
	ReadSet                     string         `json:"read_set" gorm:"type:json"`
	WriteSet                    string         `json:"write_set" gorm:"type:json"`
	ReadKeyList                 pq.StringArray `json:"read_key_list" gorm:"type:text[]"`
	WriteKeyList                pq.StringArray `json:"write_key_list" gorm:"type:text[]"`
	ChaincodeFunction           string         `json:"chaincode_function" gorm:"type:varchar(256)"`
	ChaincodeFunctionParameters string         `json:"chaincode_function_parameters" gorm:"type:varchar(256)"`
	CreateMspId                 string         `json:"create_msp_id" gorm:"type:varchar(256)"`
	EndorserMspId               string         `json:"endorser_msp_id" gorm:"type:varchar(256)"`
	EndorserSignature           string         `json:"endorser_signature" gorm:"type:text"`
	CreatedAt                   time.Time      `json:"created_at"`
	UpdatedAt                   time.Time      `json:"updated_at"`
}

func (t *Transaction) TableName() string {
	return "transaction"
}

type TransactionSearch struct {
	ChannelId int    `json:"channel_id"`
	StartTime string `json:"start_time"`
	BlockNum  uint64 `json:"block_num"`
	EndTime   string `json:"end_time"`
	PageNum   int    `json:"page_num"`
	PageCount int    `json:"page_count"`
	SortBy    string `json:"sort_by"`
	CreateMSP string `json:"create_msp"`
}

func InsertTx(session *gorm.DB, tx *Transaction) error {
	logger.Debug("insert one transaction")
	if session == nil {
		session = db
	}
	return session.Create(tx).Error
}

func GetCreateMspFromTxs(channelId int) ([]string, error) {
	var createMSPs []string
	err := db.Model(&Transaction{}).Select("create_msp_id").Where("channel_id = ?", channelId).Group("create_msp_id").Find(&createMSPs).Error
	if err != nil {
		logger.WithField("error", err).Error("failed to get craeteMSP from txs")
		return nil, errors.Wrap(err, "failed to get craeteMSP from txs")
	}

	return createMSPs, nil
}

func GetTxByTxHash(channelId int, txHash string) (*Transaction, error) {
	tx := new(Transaction)

	err := db.Where("channel_id = ? and tx_hash = ?", channelId, txHash).First(tx).Error
	if err != nil {
		logger.WithField("error", err).Error("failed to get tx by tx hash")
		return nil, err
	}

	return tx, nil
}

func GetTxCountByBlockNum(channelId int, blockNum uint64) (int64, error) {
	var count int64
	err := db.Table("transaction").Select("count(id)").
		Where("channel_id = ? and block_num = ?", channelId, blockNum).
		Find(&count).Error
	if err != nil {
		logger.WithField("error", err).Error("failed to get the count of tx by block num")
		return 0, err
	}

	return count, nil
}

func GetTxsByBlockNum(search TransactionSearch) ([]Transaction, error) {
	if search.PageNum == 0 {
		search.PageNum = 1
	}
	if search.PageCount == 0 {
		search.PageCount = 15
	}
	if search.SortBy == "" {
		search.SortBy = "desc"
	}
	var txs []Transaction
	err := db.Where("channel_id = ? and block_num = ?", search.ChannelId, search.BlockNum).Offset((search.PageNum - 1) * search.PageCount).Limit(search.PageCount).
		Order("tx_num " + search.SortBy).Find(&txs).Error
	if err != nil {
		logger.WithField("error", err).Error("failed to get tx by block number")
		return nil, err
	}

	return txs, nil
}

func GetTxsByTimeRange(search TransactionSearch) ([]Transaction, error) {
	var txs []Transaction
	if search.PageNum == 0 {
		search.PageNum = 1
	}
	if search.PageCount == 0 {
		search.PageCount = 15
	}
	if search.SortBy == "" {
		search.SortBy = "desc"
	}

	tx := db.Where("channel_id = ? ", search.ChannelId).Order("block_num " + search.SortBy).Order("tx_num " + search.SortBy).
		Limit(search.PageCount).Offset((search.PageNum - 1) * search.PageCount)

	if search.StartTime != "" {
		tx = tx.Where("created_at >= ?", search.StartTime)
	}
	if search.EndTime != "" {
		tx = tx.Where("created_at <= ?", search.EndTime)
	}
	if search.CreateMSP != "" {
		tx = tx.Where("create_msp_id = ?", search.CreateMSP)
	}

	err := tx.Find(&txs).Error
	if err != nil {
		logger.WithField("error", err).Error("failed to get txs by time range")
		return nil, err
	}

	return txs, nil
}

func GetTxCountByTimeRange(search TransactionSearch) (int64, error) {
	var count int64

	tx := db.Table("transaction").Select("count(id)").Where("channel_id = ?", search.ChannelId)

	if search.StartTime != "" {
		tx = tx.Where("created_at >= ?", search.StartTime)
	}
	if search.EndTime != "" {
		tx = tx.Where("created_at <= ?", search.EndTime)
	}
	if search.CreateMSP != "" {
		tx = tx.Where("create_msp_id = ?", search.CreateMSP)
	}

	err := tx.Find(&count).Error
	if err != nil {
		logger.WithField("error", err).Error("failed to get tx count by time range")
		return 0, err
	}

	return count, nil
}

func CountTxsByOrg() ([]map[string]interface{}, error) {
	datas := make([]map[string]interface{}, 0)

	err := db.Table("transaction").Select("create_msp_id as name, count(id) as value").Group("create_msp_id").Find(&datas).Error
	if err != nil {
		logger.WithField("error", err).Error("failed to get count of txs by organizations")
		return nil, err
	}

	return datas, nil
}

func CountTxsByWeek(search TransactionSearch) ([]int64, error) {
	var week []int64

	err := db.Debug().Raw("SELECT COALESCE(b.c,0) from (select to_char(b, 'YYYY-MM-DD') as time from generate_series(to_timestamp(?, 'YYYY-MM-DD'), to_timestamp(?, 'YYYY-MM-DD'), '1 day') AS b GROUP BY time order by time) as a full outer join (select count(t2.id) as c, to_char(t2.created_at,'YYYY-MM-DD') as time from transaction t2 where t2.channel_id = ? group by time ) as b on a.time=b.time order by a.time",
		search.StartTime, search.EndTime, search.ChannelId, search.EndTime, search.StartTime).Find(&week).Error
	if err != nil {
		logger.WithField("error", err).Error("failed to get count of txs by week")
		return nil, err
	}
	return week, nil
}

func CountTxsByDay(search TransactionSearch) ([]int64, error) {
	var day []int64
	start := search.StartTime + " 00"
	end := search.StartTime + " 23"
	err := db.Debug().Raw("SELECT COALESCE(b.c,0) from (select to_char(b, 'YYYY-MM-DD hh24') as time from generate_series(to_timestamp(?, 'YYYY-MM-DD hh24'), to_timestamp(?, 'YYYY-MM-DD hh24'), '1 hour') AS b GROUP BY time order by time) as a full outer join (select count(t2.id) as c, to_char(t2.created_at,'YYYY-MM-DD hh24') as time from transaction t2 where t2.channel_id = ? group by time ) as b on a.time=b.time where a.time is not null order by a.time",
		start, end, search.ChannelId).Find(&day).Error
	if err != nil {
		logger.WithField("error", err).Error("failed to get count of txs by day")
		return nil, err
	}
	return day, nil
}

func CountTxsByMonth(search TransactionSearch) ([]int64, error) {
	var month []int64
	err := db.Raw("SELECT COALESCE(b.c,0) from (select to_char(b, 'YYYY-MM-DD') as time from generate_series(to_timestamp(?, 'YYYY-MM-DD'), to_timestamp(?, 'YYYY-MM-DD'), '1 day') AS b GROUP BY time order by time) as a full outer join (select count(t2.id) as c, to_char(t2.created_at,'YYYY-MM-DD') as time from transaction t2 where t2.channel_id = ? group by time) as b on a.time=b.time order by a.time\n",
		search.StartTime, search.EndTime, search.ChannelId).Find(&month).Error
	if err != nil {
		logger.WithField("error", err).Error("failed to get count of txs by day")
		return nil, err
	}
	return month, nil
}
